### PR TITLE
NH-4052 - Collect schema validation exceptions

### PR DESCRIPTION
**Pull request by DMDTT**
[NH-4052](https://nhibernate.jira.com/browse/NH-4052) - Collect schema validation exceptions

+ add list of exceptions to HibernateException
+ create HibernateException on validation fail and add to the exception list
+ return list of exceptions to throw one common HibernateException to be able to evaluate it on application side

### CODE REVIEW

File: src/NHibernate/Mapping/Table.cs [hunk 1]
```diff
@@ -972,30 +972,34 @@ public override string ToString()
...
 			foreach (Column column in iter)
 			{
 				IColumnMetadata columnInfo = tableInfo.GetColumnMetadata(column.Name);
 
 				if (columnInfo == null)
 				{
-					throw new HibernateException(string.Format("Missing column: {0} in {1}", column.Name,
-															   dialect.Qualify(tableInfo.Catalog, tableInfo.Schema, tableInfo.Name)));
+					exceptions.Add(new HibernateException(string.Format("Missing column: {0} in {1}", column.Name,
+															   dialect.Qualify(tableInfo.Catalog, tableInfo.Schema, tableInfo.Name))));
```
Comment (fredericDelaporte) [inline src/NHibernate/Mapping/Table.cs]: Here lies a bug due to the change. A throw stops code execution, protecting in this case the code from reaching a null ref condition. You should make sure that everywhere you have replaced a throw by just an add in a list, the code following that will support the state for which an exception was supposed to be thrown or will be appropriately bypassed.

File: src/NHibernate/HibernateException.cs [hunk 2]
```diff
@@ -12,6 +13,7 @@ namespace NHibernate
 	[Serializable]
 	public class HibernateException : Exception
 	{
+		public List<HibernateException> Exceptions { get; }
```
Comment (fredericDelaporte) [inline src/NHibernate/HibernateException.cs]: Better follow usual conventions, like the one of the standard [`AggregateException`](https://docs.microsoft.com/en-us/dotnet/api/system.aggregateexception): name this `InnerExceptions`, and expose this as a `ReadOnlyCollection<Exception>`.

File: src/NHibernate/HibernateException.cs [hunk 3]
```diff
@@ -12,6 +13,7 @@ namespace NHibernate
 	[Serializable]
 	public class HibernateException : Exception
```
Comment (fredericDelaporte) [inline src/NHibernate/HibernateException.cs]: I am not for adding this to the hibernate exception base class. If we add that, it should be on a specialized class, like adding `class HibernateAggregateException : HibernateException`.

**Comment by DMDTT** [comment]
@fredericDelaporte Thanks for your review.
Keeping the validation errors as a readonlycollection seems a more apropriate way then keeping one string.

I updated the code

File: src/NHibernate.Test/App.config [hunk 4]
```diff
@@ -7,7 +7,7 @@
 	</configSections>
 
 	<connectionStrings>
-		<add name="TestConnectionString" connectionString="Server=localhost\sqlexpress;Database=nhibernate;Integrated Security=SSPI" />
+		<add name="TestConnectionString" connectionString="Server=localhost;Database=nhibernate;Integrated Security=SSPI" />
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/App.config]: Please revert. You should not need to change that. Use ShowBuildMenu and setup your test configuration with it.

File: src/NHibernate/HibernateException.cs [hunk 5]
```diff
@@ -1,4 +1,5 @@
 using System;
+using System.Collections.Generic;
```
Comment (fredericDelaporte) [inline src/NHibernate/HibernateException.cs]: Change now unneeded, please revert too.

File: src/NHibernate.TestDatabaseSetup/App.config [hunk 6]
```diff
@@ -9,7 +9,7 @@
     <session-factory name="NHibernate.TestDatabaseSetup">
 			<property name="connection.driver_class">NHibernate.Driver.SqlClientDriver</property>
 			<property name="connection.connection_string">
-        Server=.\SQLExpress;initial catalog=master;Integrated Security=SSPI
+        Server=localhost;initial catalog=master;Integrated Security=SSPI
```
Comment (fredericDelaporte) [inline src/NHibernate.TestDatabaseSetup/App.config]: Please revert too.

File: src/NHibernate/SchemaValidationException.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
+﻿using System.Collections.Generic;
+using System.Collections.ObjectModel;
+
+namespace NHibernate
+{
+	public class SchemaValidationException : HibernateException
+	{
+		public ReadOnlyCollection<string> ValidationErrors { get; }
+
+		public SchemaValidationException(string msg, IList<string> validationErrors) : base(msg)
```
Comment (fredericDelaporte) [inline src/NHibernate/SchemaValidationException.cs]: `IEnumerable<string>` should be enough.

File: src/NHibernate.Test/NHSpecificTest/NH1274ExportExclude/NH1274ExportExcludeFixture.cs [hunk 8]
```diff
@@ -7,6 +7,7 @@
 using NHibernate.Criterion;
 using NHibernate.Engine;
 using NHibernate.Tool.hbm2ddl;
+using NHibernate.Util;
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/NH1274ExportExclude/NH1274ExportExcludeFixture.cs]: We wish to remove NHibernate collection custom extensions, so please do not add more usages of them.

File: src/NHibernate.Test/NHSpecificTest/NH1274ExportExclude/NH1274ExportExcludeFixture.cs [hunk 9]
```diff
@@ -84,9 +85,10 @@ public void SchemaExport_Validate_CausesValidateException()
 			{
 				validator.Validate();
 			}
-			catch (HibernateException he)
+			catch (SchemaValidationException e)
 			{
-				Assert.IsTrue(he.Message.Contains("Home_Validate"));
+				Assert.IsTrue(e.Message.Contains("Schema validation failed: see list of validation errors"));
+				Assert.IsTrue(e.ValidationErrors.Contains("Missing table: Home_Validate"));
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/NH1274ExportExclude/NH1274ExportExcludeFixture.cs]: Better use Fluent syntax: `Assert.That`. NUnit does no more add features to the others, and may obsolete them.
By the way, just having "expected true but was false" as an error message while there is more than one test on `true` is not practical : better at least add an explicit error message.

Here the whole thing could be simplified too:

```C#
Assert.That(() => validator.Validate(),
    Throws.TypeOf<SchemaValidationException>()
        .And.Property("ValidationErrors").Contain("Missing table: Home_Validate"));
```
(Checking the type and the error list without checking the exception message seems enough to me, the type is already sufficiently specialized.)

File: src/NHibernate.Test/Tools/hbm2ddl/SchemaValidator/SchemaValidateFixture.cs [hunk 10]
```diff
@@ -60,9 +61,10 @@ public void ShouldNotVerifyModifiedTable()
 			{
 				validatorV2.Validate();
 			}
-			catch (HibernateException e)
+			catch (SchemaValidationException e)
 			{
-				Assert.That(e.Message, Does.StartWith("Missing column: Name"));
+				Assert.That(e.Message, Does.StartWith("Schema validation failed: see list of validation errors"));
+				Assert.That(e.ValidationErrors[0], Does.StartWith("Missing column: Name"));
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/Tools/hbm2ddl/SchemaValidator/SchemaValidateFixture.cs]: Better but could be simplified as previous assert.

File: src/NHibernate/SchemaValidationException.cs (hunk 7 above)
Comment (hazzik) [reply src/NHibernate/SchemaValidationException.cs]: ReadOnlyCollection (next line) requires `IList<T>`

**Comment by hazzik** [comment]
Mentioned issues resolved. Squashed, rebased and Async tests regenerated.

File: src/NHibernate.Test/Async/Tools/hbm2ddl/SchemaValidator/SchemaValidateFixture.cs [hunk 11]
```diff
@@ -67,14 +68,12 @@ public class SchemaValidateFixtureAsync
...
-			catch (HibernateException e)
-			{
-				Assert.That(e.Message, Does.StartWith("Missing column: Name"));
-			}
+
+			Assert.That(
+				() => validatorV2.ValidateAsync(),
+				Throws.TypeOf<SchemaValidationException>()
+				      .And.Message.EqualTo("Schema validation failed: see list of validation errors")
+				      .And.Property("ValidationErrors").Contains("Missing column: Name in nhibernate.dbo.Version"));
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/Async/Tools/hbm2ddl/SchemaValidator/SchemaValidateFixture.cs]: As showcased by failing tests, this column naming is specific to SQL Server and causes the test to fail for other db.

**Comment by hazzik** [comment]
I'm not doing it right now. So yes, if you wish

**Comment by DMDTT** [comment]
@fredericDelaporte @hazzik 
Thanks for your work guys, 
I didn't manage it to work on it in the last month, sry for that.

But nethertheless the work of yours is great. Thanks again.

**Pull request merged**